homeassistant/components/melcloud/climate.py

```diff
@@ -0,0 +1,200 @@
+        """
+        return True
+
+    async def async_update(self):
+        """Update state from MELCloud."""
+        await self._api.async_update()
+
+    @property
+    def device_info(self):
+        """Return a device description for device registry."""
+        return self._api.device_info
+
+    @property
+    def state(self) -> str:
+        """Return the current state."""
+        return self.hvac_mode
+
+    @property
+    def precision(self) -> float:
+        """Return the precision of the system."""
+        if self.hass.config.units.temperature_unit == TEMP_CELSIUS:
+            return PRECISION_TENTHS
+        return PRECISION_WHOLE
+
+    @property
+    def temperature_unit(self) -> str:
+        """Return the unit of measurement used by the platform."""
+        return TEMP_UNIT_LOOKUP.get(self._api.device.temp_unit, TEMP_CELSIUS)
+
+    @property
+    def hvac_mode(self) -> str:
+        """Return hvac operation ie. heat, cool mode."""
+        mode = self._api.device.operation_mode
+        if not self._api.device.power or mode is None:
+            return HVAC_MODE_OFF
+        return HVAC_MODE_LOOKUP.get(mode)
+
+    async def async_set_hvac_mode(self, hvac_mode: str) -> None:
+        """Set new target hvac mode."""
+        if hvac_mode == HVAC_MODE_OFF:
+            await self._api.device.set({"power": False})
+            return
+
+        operation_mode = HVAC_MODE_REVERSE_LOOKUP.get(hvac_mode, None)
+        if operation_mode is None:
+            raise ValueError(f"Invalid hvac_mode [{hvac_mode}]")
+
+        props = {"operation_mode": operation_mode}
+        if self.hvac_mode == HVAC_MODE_OFF:
+            props["power"] = True
+        await self._api.device.set(props)
+
+    @property
+    def hvac_modes(self) -> List[str]:
+        """Return the list of available hvac operation modes."""
+        return [HVAC_MODE_OFF] + list(
+            map(HVAC_MODE_LOOKUP.get, self._api.device.operation_modes())
+        )
+
+    @property
+    def current_temperature(self) -> Optional[float]:
+        """Return the current temperature."""
+        return self._api.device.temperature
+
+    @property
+    def target_temperature(self) -> Optional[float]:
+        """Return the temperature we try to reach."""
+        return self._api.device.target_temperature
+
+    async def async_set_temperature(self, **kwargs) -> None:
+        """Set new target temperature."""
+        await self._api.device.set(
+            {"target_temperature": kwargs.get("temperature", self.target_temperature)}
+        )
+
+    @property
+    def target_temperature_step(self) -> Optional[float]:
+        """Return the supported step of target temperature."""
+        return self._api.device.target_temperature_step
+
+    @property
+    def fan_mode(self) -> Optional[str]:
+        """Return the fan setting."""
+        speed = self._api.device.fan_speed
+        if speed is None:
+            return None
+        return speed.replace("-", " ")
```

Comment:
There's a supposedly dynamic amount of fan speeds available in these devices. The backend lib doing this `speed-%d` mapping for them and this is just to turn that "constant-y" string to "human readable" variant. What I'm trying to say is that I think the replace belongs here, but it just sucks.

Would it be better to use the `FAN_*` fan mode constants here instead of sequentially numbered speeds? I should be possible at least, but I don't know if it has any benefits.

Reply:
Sorry, I don't know about the details, but maybe using those constants will help when it comes to consistency (think: usage via voice controls)?